File: c/802.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
bool find(int** g, int* gcs, int i, bool* pter, bool* trav) {
    if (trav[i])
        return pter[i];
    trav[i] = true;
    int n = gcs[i];
    for (int j = 0; j < n; j++)
        if (!find(g, gcs, g[i][j], pter, trav))
            return false;
    pter[i] = true;
    return true;
}

int* eventualSafeNodes(int** g, int n, int* gcs, int* rs) {
    int* res = (int*)malloc(sizeof(int) * n);
    bool* pter = (bool*)malloc(sizeof(bool) * n);
    bool* trav = (bool*)malloc(sizeof(bool) * n);
    for (int i = 0; i < n; i++) {
        pter[i] = false;
        trav[i] = false;
    }
    int l = 0;
    for (int i = 0; i < n; i++)
        if (find(g, gcs, i, pter, trav))
            res[l++] = i;
    *rs = l;
    return res;
}

File: javascript/802.js
/**
 * @param {number[][]} g
 * @return {number[]}
 */
var eventualSafeNodes = function (g) {
    var find = function (g, i, pter, trav) {
        if (trav[i])
            return pter[i];
        trav[i] = true;
        for (let j of g[i])
            if (!find(g, j, pter, trav))
                return false;
        pter[i] = true;
        return true;
    };
    let n = g.length;
    let pter = new Array(n).fill(false);
    let trav = new Array(n).fill(false);
    let res = []
    for (let i = 0; i < n; i++)
        if (find(g, i, pter, trav))
            res.push(i);
    return res;
};
